keps/168-pending-workloads-visibility/README.md

```diff
@@ -0,0 +1,349 @@
+the position of my workload in the cluster queue, so that I can estimate the
+ETA of my workload.
+
+I would like to be able to get this information by inspecting the cluster queue
+status.
+-->
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+#### Time and memory consuming computation of top pending workloads
+
+Currently, we organize the cluster queue as a heap, thus only the top of the
+heap is readily available. If we want to get to know the top N pending workloads
+we may need to copy the heap, sort and select N, which is time and memory
+consuming.
+
+In order to mitigate this risk we may need to migrate from using a heap to
+red-black trees.
+
+#### Too large local queue object
+
+As the number of pending workloads is arbitrarily large there is a risk that the
+status information about the workloads may exceed the etcd limit of 1.5Mi on
+object size. In order to allow feature extensions of the structure we
+should assume not more that 500Ki is used.
+
+In order to mitigate this risk we put a constraint on the number of exposed
+pending workloads. We limit the number to 1000.
+
+#### Increased number of localqueue updates
+
+As we put the global cluster queue position into the list of top pending workloads
+changes to one local queue may trigger updates in another local queue connected
+to the same cluster queue, due to status changes in the structure.
+
+In order to mitigate this risk we need to consider:
+1. batching updates to local queues, by a batch period, similarly as Job status
+   updates are batched in Kubernetes.
+2. allow to control the number of N workloads which expose their position in the
+   cluster queue. Setting low value for the limit would reduce the change of
+   changes in one local queue to trigger an update in another local queue.
+
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Local Queue API
+
+```golang
+// PendingWorkload contains the information identifying a pending workload in
+// the local queue.
+type PendingWorkload struct {
+	// Name indicates the name of the pending workload.
+	Name string
+
+	// Position indicates the position of the workload among all pending
+	// workloads in the cluster queue.
+	Position *int32
+}
+
+type PendingWorkloadsStatus struct {
+	// TopList contains the list of top pending workloads.
+	// +listType=map
+	// +listMapKey=name
+	// +listMapKey=namespace
+	// +optional
+	TopList []PendingWorkload
```

Comment:
`headWorkloads`? I'm open to suggestions.

Reply:
Yeah, I was thinking about `head`, but it seems it might suggest a single element, so then I though `top` is better :), but open to other suggestions as well.

Reply:
Let's go with just head, adjusted.